### PR TITLE
*: manifest-list -> image-index

### PR DESCRIPTION
**Pull request by vbatts**
per discussion around https://github.com/opencontainers/image-spec/pull/533

Signed-off-by: Vincent Batts <vbatts@hashbangbash.com>

### CODE REVIEW

**Comment by stevvooe** [comment]
LGTM

:+1: :+1: :+1: :+1: :+1: :+1: :+1: :+1:

File: image-layout.md [hunk 1]
```diff
@@ -5,7 +5,7 @@ This layout MAY be used in a variety of different transport mechanisms: archive
 
 Given an image layout and a ref, a tool can create an [OCI Runtime Specification bundle](https://github.com/opencontainers/runtime-spec/blob/v1.0.0-rc2/bundle.md) by:
 
-* Following the ref to find a [manifest](manifest.md#image-manifest), possibly via a [manifest list](manifest-list.md#manifest-list)
+* Following the ref to find a [manifest](manifest.md#image-manifest), possibly via a [image index](image-index.md#manifest-list)
```
Comment (wking) [inline image-layout.md]: “a image index” → “an image index”?

And do we really need to call this an “image index”?  Can't it just be an “index”?  There doesn't seem to be much specific to images about an array of descriptors.

File: schema/image-index-schema.json [hunk 2]
```diff
@@ -1,12 +1,12 @@
 {
   "description": "OpenContainer Image Manifest List Specification",
   "$schema": "http://json-schema.org/draft-04/schema#",
-  "id": "https://opencontainers.org/schema/image/manifest-list",
+  "id": "https://opencontainers.org/schema/image/image-index",
```
Comment (wking) [inline schema/image-index-schema.json]: Here's [another place][1] where “image index” seems over specific.  Can we follow your  `…vnd.oci.image.index…` media type (which does not double up on `index` and use `…/image/index` here (and the other `id`s you touch)?

[1]: https://github.com/opencontainers/image-spec/pull/546#discussion_r99223558

File: image-layout.md (hunk 1 above)
Comment (vbatts) [inline image-layout.md]: this is like the Image Layout, no?

Comment (wking) [inline image-layout.md]: > this is like the Image Layout, no?

Yeah, we should probably rename that too ;).

The only thing particularly image-specific about image-spec is the config format, and when we have sufficient context, we refer to that as “[a configuration][1]”.  The full name (for when you don't have sufficient context) is “[OCI Image Configuration][2]”.  I think that's a good pattern to follow, and it would mean we'd have something like the following formats:

| Current | Short | Long |
| -- | -- | -- |
| image layout | directory storage | OCI Image Directory Storage |
| manifest list | index | OCI Image Index |
| manifest | manifest | OCI Image Manifest |
| configuration | configuration | OCI Image Configuration |
| layer (and "[layer filesystem changeset](https://github.com/opencontainers/image-spec/blob/v1.0.0-rc4/layer.md#image-layer-filesystem-changeset)”) | layer | OCI Image Layer |

[1]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/spec.md#L3
[2]: https://github.com/opencontainers/image-spec/blob/v1.0.0-rc4/config.md#oci-image-configuration

Comment (stevvooe) [inline image-layout.md]: > OCI Image Directory Storage

WTAF. The image layout is __NOT__ directory storage. OCI image spec _DOES NOT_ prescribe a specific directory layout for storing images.

Stop propagating your fantasy. It is confusing people and holding things up.

Comment (wking) [inline image-layout.md]: > The image layout is NOT directory storage.

Image-layout is not *storing* directories, it is [*using* a directory tree to store blobs and references][1].  You can wrap that image-layout tree up in a [tarball (or whatever you want)][2], but it's still a directory-based format.  That format stores [opaque][3], [content-addressable][4] blobs, and [generic references to them][5], so I don't see anything that is specific to images besides some [informative][6] [hints][7].

[1]: https://github.com/opencontainers/image-spec/blob/v1.0.0-rc4/image-layout.md#content
[2]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/image-layout.md#L4
[3]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/image-layout.md#L17
[4]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/image-layout.md#L3
[5]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/image-layout.md#L60
[6]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/image-layout.md#L19
[7]: https://github.com/opencontainers/image-spec/blame/v1.0.0-rc4/image-layout.md#L65

Comment (stevvooe) [inline image-layout.md]: > OCI Image Directory Storage

To everyone else in the world, this does not mean a tar.

File: spec.md [hunk 3]
```diff
@@ -50,7 +50,7 @@ The [OCI Image Media Types](media-types.md) document is a starting point to unde
 
 The high-level components of the spec include:
 
-* An archival format for container images, consisting of an [image manifest](manifest.md), a [manifest list](manifest-list.md) (optional), an [image layout](image-layout.md), a set of [filesystem layers](layer.md), and [image configuration](config.md) (base OCI layer)
+* An archival format for container images, consisting of an [image manifest](manifest.md), a [image index](image-index.md) (optional), an [image layout](image-layout.md), a set of [filesystem layers](layer.md), and [image configuration](config.md) (base OCI layer)
```
Comment (AkihiroSuda) [inline spec.md]: nit: a image -> an image

File: image-layout.md (hunk 1 above)
Comment (vbatts) [inline image-layout.md]: @wking i'll take a look for where it may stutter a bit, but your table is a excessive . :grimacing:

File: spec.md (hunk 3 above)
Comment (vbatts) [inline spec.md]: fixed

**Comment by vbatts** [comment]
Updated. PTAL.

**Comment by stevvooe** [comment]
@vbatts CI is failing.

**Comment by vbatts** [comment]
fixed. `dot` wasn't on travis, and the image needed to be updated.

**Comment by stevvooe** [comment]
LGTM

File: image-layout.md (hunk 1 above)
Comment (stevvooe) [inline image-layout.md]: @vbatts @wking I had a visceral rejection to "Directory Storage".

I would be fine with this table if we called it `OCI Image Archive`.

It is out of scope of this issue. Let's get this one merged for the media type change.

File: spec.md [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Open Container Initiative
 ## Image Format Specification
 
-This specification defines an OCI Image, consisting of a [manifest](manifest.md), a [manifest list](manifest-list.md) (optional), a set of [filesystem layers](layer.md), and a [configuration](config.md).
+This specification defines an OCI Image, consisting of a [manifest](manifest.md), a [image index](image-index.md) (optional), a set of [filesystem layers](layer.md), and a [configuration](config.md).
```
Comment (wking) [inline spec.md]: “a image” → “an image”.

File: manifest.md [hunk 5]
```diff
@@ -3,15 +3,15 @@
 There are three main goals of the Image Manifest Specification.
 The first goal is content-addressable images, by supporting an image model where the image's configuration can be hashed to generate a unique ID for the image and its components.
 The second goal is to allow multi-architecture images, through a "fat manifest" which references image manifests for platform-specific versions of an image.
-In OCI, this is codified in a [Manifest List](manifest-list.md).
+In OCI, this is codified in a [Image Index](image-index.md).
```
Comment (wking) [inline manifest.md]: “a Image” → “an Image” (and this “a i…” → “an i…” shows up in a few other places too).

File: media-types.md [hunk 6]
```diff
@@ -62,6 +62,6 @@ The following figure shows how the above media types reference each other:
 ![](img/media-types.png)
 
 [Descriptors](descriptor.md) are used for all references.
-The manifest list being a "fat manifest" references one or more image manifests per target platform. An image manifest references exactly one target configuration and possibly many layers.
+The image-index being a "fat manifest" references one or more image manifests per target platform. An image manifest references exactly one target configuration and possibly many layers.
```
Comment (wking) [inline media-types.md]: You don't hyphenate “image index” in most of the other places where you mention it.  I don't care if you hyphenate it or not, but it's probably worth picking one way and staying consistent.

File: schema/fs.go [hunk 7]
```diff
@@ -202,227 +202,310 @@ func _escFSMustString(useLocal bool, name string) string {
 
 var _escData = map[string]*_escFile{
 
-	"/.content-descriptor.json.un~": {
-		local:   ".content-descriptor.json.un~",
-		size:    1548,
-		modtime: 1485467092,
+	"/.fs.go.un~": {
```
Comment (wking) [inline schema/fs.go]: More local files.  I think you should land #538 (I've added a comment reproducing my successfully ignoring local files like this to that PR) or remove your local editor files before rebuilding `fs.go`.

**Comment by wking** [comment]
Also, still some dangling references to clean up:

```
$ git grep -i manifest.list origin/pr/546
origin/pr/546:annotations.md:This specification defines the following annotation keys, intended for but not limited to manifest list and image manifest authors:
origin/pr/546:image-index.md:The manifest list is a higher-level manifest which points to specific [image manifests](manifest.md) for one or more platforms.
origin/pr/546:image-index.md:While the use of a manifest list is OPTIONAL for image providers, image consumers SHOULD be prepared to process them.
origin/pr/546:image-index.md:    Manifest lists concerned with portability SHOULD use one of the above media types.
origin/pr/546:image-index.md:        Manifest lists SHOULD use, and implementations SHOULD understand, values [supported by runtime-spec's `platform.arch`][runtime-platform2].
origin/pr/546:image-index.md:        Manifest lists SHOULD use, and implementations SHOULD understand, values [supported by runtime-spec's `platform.os`][runtime-platform2].
origin/pr/546:media-types.md:- [application/vnd.docker.distribution.manifest.list.v2+json](https://github.com/docker/distribution/blob/master/docs/spec/manifest-v2-2.md#manifest-list) - mediaType is different
origin/pr/546:schema/image-index-schema.json:  "description": "OpenContainer Image Manifest List Specification",
origin/pr/546:schema/manifest_backwards_compatibility_test.go:  "application/vnd.docker.distribution.manifest.list.v2+json": v1.MediaTypeImageIndex,
origin/pr/546:schema/manifest_backwards_compatibility_test.go:   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
origin/pr/546:specs-go/v1/image_index.go:       // Annotations contains arbitrary metadata for the manifest list.
```

The `application/vnd.docker.distribution.manifest.list.v2+json` entries are fine, but the others should be updated.

File: image-index.md [hunk 8]
```diff
@@ -1,12 +1,12 @@
-# OCI Image Manifest List Specification
+# OCI Image Index Specification
 
 The manifest list is a higher-level manifest which points to specific [image manifests](manifest.md) for one or more platforms.
```
Comment (coolljt0725) [inline image-index.md]: should this `manifest list` also change to `image index `?

File: image-index.md [hunk 9]
```diff
@@ -1,12 +1,12 @@
...
 While the use of a manifest list is OPTIONAL for image providers, image consumers SHOULD be prepared to process them.
```
Comment (coolljt0725) [inline image-index.md]: same here `While the use of a manifest list is OPTIONAL` 
and also https://github.com/opencontainers/image-spec/pull/546/files#diff-08e132d566e81340ee3533c41567d487R54 and https://github.com/opencontainers/image-spec/pull/546/files#diff-08e132d566e81340ee3533c41567d487R49

File: specs-go/v1/image_index.go [hunk 10]
```diff
@@ -50,9 +50,9 @@ type ManifestDescriptor struct {
 	Platform Platform `json:"platform"`
 }
 
-// ManifestList references manifests for various platforms.
-// This structure provides `application/vnd.oci.image.manifest.list.v1+json` mediatype when marshalled to JSON.
-type ManifestList struct {
+// ImageIndex references manifests for various platforms.
+// This structure provides `application/vnd.oci.image.index.v1+json` mediatype when marshalled to JSON.
+type ImageIndex struct {
```
Comment (coolljt0725) [inline specs-go/v1/image_index.go]: comment of  Annotations  `// Annotations contains arbitrary metadata for the manifest list.` below should change to `manifest list` to `image index`

**Comment by coolljt0725** [comment]
oh, sorry for the noise :), seems @wking  has point out these reference that's need to be clean up in https://github.com/opencontainers/image-spec/pull/546#issuecomment-279805769

**Comment by jonboulle** [comment]
seems like this needs a rebase and a few more fixes from the last couple reviews (at a glance I still see several manifest list references too)

File: image-layout.md [hunk 11]
```diff
@@ -162,12 +162,12 @@ Those tags will often be represented in an image-layout repository with matching
...
 ### Index Example
 
-```json,title=Manifest%20List&mediatype=application/vnd.oci.image.manifest.list.v1%2Bjson
+```json,title=Manifest%20List&mediatype=application/vnd.oci.image.index.v1%2Bjson
 {
   "schemaVersion": 2,
   "manifests": [
     {
-      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "mediaType": "application/vnd.oci.image.index.v1+json",
```
Comment (gao-feng) [inline image-layout.md]: why this change?

File: image-layout.md [hunk 12]
```diff
@@ -162,12 +162,12 @@ Those tags will often be represented in an image-layout repository with matching
...
       "size": 7143,
       "digest": "sha256:0228f90e926ba6b96e4f39cf294b2586d38fbb5a1e385c05cd1ee40ea54fe7fd",
       "annotations": {
```
Comment (gao-feng) [inline image-layout.md]: Is `org.opencontainers.ref.name` in `annotations` valid for image index?

Comment (vbatts) [inline image-layout.md]: yes. Just like a file `./refs/stable` could be a descriptor of a manifest-list, this named reference points to a now image-index (previously manifest-list). I only changed this example to demo an image-index, rather than only a manifest.

**Comment by vbatts** [comment]
rebased and fixed. PTAL

File: considerations.md [hunk 13]
```diff
@@ -1,6 +1,6 @@
 # Extensibility
 
-Implementations that are reading/processing [manifests](manifest.md) or [manifest lists](manifest-list.md) MUST NOT generate an error if they encounter an unknown property.
+Implementations that are reading/processing [manifests](manifest.md) or [image index](image-index.md) MUST NOT generate an error if they encounter an unknown property.
```
Comment (wking) [inline considerations.md]: "image indexes" or "the image index"?

File: annotations.md [hunk 14]
```diff
@@ -15,7 +15,7 @@ Consumers MUST NOT generate an error if they encounter an unknown annotation key
 
 ## Pre-Defined Annotation Keys
 
-This specification defines the following annotation keys, intended for but not limited to manifest list and image manifest authors:
+This specification defines the following annotation keys, intended for but not limited to [image-index](image-index.md) and image [manifest](manifest.md) authors:
```
Comment (wking) [inline annotations.md]: Hyphenated "image-index" here (you seem to prefer the unhyphenated "image index").

File: image-index.md [hunk 15]
```diff
@@ -1,12 +1,12 @@
-# OCI Image Manifest List Specification
+# OCI Image Index Specification
 
-The manifest list is a higher-level manifest which points to specific [image manifests](manifest.md) for one or more platforms.
-While the use of a manifest list is OPTIONAL for image providers, image consumers SHOULD be prepared to process them.
+The Image Index is a higher-level manifest which points to specific [image manifests](manifest.md), ideal for one or more platforms.
```
Comment (wking) [inline image-index.md]: Title-case "Image Index" here (you seem to prefer the sentence-case "image index).

File: image-index.md [hunk 16]
```diff
@@ -34,7 +34,7 @@ For the media type(s) that this document is compatible with, see the [matrix][ma
 
     - [`application/vnd.oci.image.manifest.v1+json`](manifest.md)
 
-    Manifest lists concerned with portability SHOULD use one of the above media types.
+    Image Indexes are concerned with portability SHOULD use one of the above media types.
```
Comment (jonboulle) [inline image-index.md]: s/are concerned/concerned/

File: media-types.md (hunk 6 above)
Comment (jonboulle) [inline media-types.md]: prefer without

File: image-index.md (hunk 16 above)
Comment (vbatts) [inline image-index.md]: fixed

File: image-index.md (hunk 15 above)
Comment (vbatts) [inline image-index.md]: fixed

File: considerations.md (hunk 13 above)
Comment (vbatts) [inline considerations.md]: fixed

File: annotations.md (hunk 14 above)
Comment (vbatts) [inline annotations.md]: fixed

**Comment by vbatts** [comment]
updated PTAL

**Comment by philips** [comment]
LGTM

File: image-index.md [hunk 17]
```diff
@@ -1,12 +1,12 @@
-# OCI Image Manifest List Specification
+# OCI Image Index Specification
 
-The manifest list is a higher-level manifest which points to specific [image manifests](manifest.md) for one or more platforms.
-While the use of a manifest list is OPTIONAL for image providers, image consumers SHOULD be prepared to process them.
+The Image index is a higher-level manifest which points to specific [image manifests](manifest.md), ideal for one or more platforms.
```
Comment (wking) [inline image-index.md]: I'm not sure if we're punting copy-edits to a follow-up commit or not, but “Image” → “image” (either in this PR or in the follow-up copy-edit).

File: img/media-types.dot [hunk 18]
```diff
@@ -1,12 +1,12 @@
 digraph G {
   {
-    manifestList [shape=note, label="Manifest list\n<<optional>>\napplication/vnd.oci.image.manifest.list.v1+json"]
+    imageIndex [shape=note, label="Image Index<<optional>>\napplication/vnd.oci.image.index.v1+json"]
```
Comment (wking) [inline img/media-types.dot]: This dropped a newline with “Manifest list\n” → “Image Index”.  I don't know if that was intentional or not, although my preference would be to keep the newline.

File: schema/fs.go [hunk 19]
```diff
@@ -379,735 +452,9107 @@ hR+PkX5zgcX30ucJvpgCXqksmwP73wn8U2qZo81xwnEy+6yyZ6Vexsl79az+DgAA//850zAaggUAAA==
 
 	"/fs.go": {
```
Comment (wking) [inline schema/fs.go]: Ahh, recursive `http.FileSystem`!  And not new to this PR either :p.  In fact, it looks like it goes all the way back to #65.

File: schema/image-manifest-schema.json [hunk 20]
```diff
@@ -22,7 +22,7 @@
       }
     },
     "annotations": {
-      "id": "https://opencontainers.org/schema/image/manifest-list/annotations",
+      "id": "https://opencontainers.org/schema/image/manifest/annotations",
```
Comment (wking) [inline schema/image-manifest-schema.json]: This is also [in flight][1] with #484.  I guess whichever way gets the fix in first… ;)

[1]: https://github.com/opencontainers/image-spec/pull/484/files#diff-55ff3e2e7005b8b807f3da93f975e3ebL25

File: specs-go/v1/manifest.go [hunk 21]
```diff
@@ -27,6 +27,6 @@ type Manifest struct {
 	// Layers is an indexed list of layers referenced by the manifest.
 	Layers []Descriptor `json:"layers"`
 
-	// Annotations contains arbitrary metadata for the manifest list.
+	// Annotations contains arbitrary metadata for the manifest.
```
Comment (wking) [inline specs-go/v1/manifest.go]: This is also [in flight][1] with #484. I guess whichever way gets the fix in first… ;)

[1]: https://github.com/opencontainers/image-spec/pull/484/files#diff-83ea0fb4782f9fe8fcc3bb3063ef1b53L30

File: image-layout.md (hunk 12 above)
Comment (gao-feng) [inline image-layout.md]: Got it, thanks!

**Comment by jbouzane** [comment]
LGTM

File: image-index.md (hunk 17 above)
Comment (vbatts) [inline image-index.md]: Well I would not mind punting copy-edits, but I've already got to rebase again.

File: img/media-types.dot (hunk 18 above)
Comment (vbatts) [inline img/media-types.dot]: will pick that up.

File: manifest.md [hunk 22]
```diff
@@ -3,15 +3,15 @@
 There are three main goals of the Image Manifest Specification.
 The first goal is content-addressable images, by supporting an image model where the image's configuration can be hashed to generate a unique ID for the image and its components.
 The second goal is to allow multi-architecture images, through a "fat manifest" which references image manifests for platform-specific versions of an image.
-In OCI, this is codified in a [Manifest List](manifest-list.md).
+In OCI, this is codified in an [Image Index](image-index.md).
```
Comment (vbatts) [inline manifest.md]: meh. I'm two ways about it.

File: schema/fs.go [hunk 23]
```diff
@@ -202,10 +202,83 @@ func _escFSMustString(useLocal bool, name string) string {
 
 var _escData = map[string]*_escFile{
 
+	"/backwards_compatibility_test.go": {
```
Comment (vbatts) [inline schema/fs.go]: ?

File: schema/fs.go (hunk 19 above)
Comment (vbatts) [inline schema/fs.go]: so, this issue is outside of this PR

**Comment by vbatts** [comment]
rebased and updated. PTAL

File: schema/fs.go (hunk 19 above)
Comment (wking) [inline schema/fs.go]: Not really, since you're supposed to be [ignoring `*.go`][1].  Your copy of `esc` just seems to ignore `-ignore=…` :p.  For example, here's [my most-recently-landed `fs.go` rebuild][2], and it only contains `*.json`.

[1]: https://github.com/vbatts/oci-image-spec/blob/ab461b048bd1c8b6077d8e96936f706a518233c2/schema/gen.go#L21
[2]: https://github.com/opencontainers/image-spec/blob/10dc107a303d3118a1c452ec59192bd20510559d/schema/fs.go

Comment (wking) [inline schema/fs.go]: And in case it helps, here's the history of the recursive `/fs.go`'s presence in master:

```
$ git log --decorate -p -G '/fs.go' origin/master -- schema/fs.go | grep '^commit \|Author\|"/fs.go"'
commit 0556a6bee5ee1514762de05fd95d7af09b38f07f (origin/pr/533)
Author: Vincent Batts <vbatts@hashbangbash.com>
+       "/fs.go": {
commit 10e6ff7101c2b0ff76e41a4a213838953e2082d8 (origin/pr/78)
Author: Jonathan Boulle <jonathanboulle@gmail.com>
-       "/fs.go": {
commit 4263cf529b30d46b99a761ca764ede70b667cdc9 (origin/pr/65)
Author: Sergiusz Urbaniak <sur@coreos.com>
+       "/fs.go": {
```

Since #533, only @xiekeyang has rebuilt `fs.go`:

```
$ git log --decorate 0556a6be..origin/master -- schema/fs.go | grep '^commit \|Author'
commit 5bf90dd1ed7a0607c7fdc6013e3841952d228d43 (origin/pr/564)
Author: xiekeyang <xiekeyang@huawei.com>
```

Who seems to have the same buggy `esc` behavior:

```
$ git cat-file -p 5bf90dd1:schema/gen.go | grep 'generate esc'
//go:generate esc -private -pkg=schema -ignore=.*go -ignore=.*swp .
$ git cat-file -p 5bf90dd1:schema/fs.go | grep '"/'
        "/config-schema.json": {
        "/config_test.go": {
```

My theory is still that you have an old, buggy copy of `esc` ;).

**Comment by vbatts** [comment]
hrm. We should have a `go get -u github.com/mjibson/esc` somewhere ...

**Comment by vbatts** [comment]
I've just updated that tool, and `make schema/fs.go` still adds it in. :-\

**Comment by stevvooe** [comment]
LGTM

**Comment by philips** [comment]
LGTM
